File: 第8章 神奇分形图/8.2 科赫雪花/科赫雪花.cpp
/*
程序：科赫雪花
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch到Python和C++》
*/
// 递归绘制科赫曲线
void draw_curve(float b, int n)
{
    if (n == 0) {         // 递归结束条件
        pen.fd(b);
        return;
    }
    
    draw_curve(b/3, n-1); // 画线段①
    
    pen.rt(60);
    draw_curve(b/3, n-1); // 画线段②
    
    pen.lt(120);
    draw_curve(b/3, n-1); // 画线段③
    
    pen.rt(60);
    draw_curve(b/3, n-1); // 画线段④
}

// 绘制科赫雪花
void draw_snowflake(float b, int n)
{
    for (int i = 0; i < 3; i++) {
        draw_curve(b, n);
        pen.lt(120);
    }
}

// 绘制一个4阶的科赫雪花
int main()
{
	pen.speed(8);
	pen.color(_black);
	pen.size(2);
	pen.up();
    pen.move(-150, -90);
    pen.angle(90);
	pen.down();
    draw_snowflake(300, 4);
    return 0;
}

File: 第8章 神奇分形图/8.2 科赫雪花/科赫雪花.py
'''
程序：科赫曲线
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch迈向Python和C++》
'''
from turtle import *

def draw_curve(b, n):
    '''递归绘制科赫曲线'''
    if n == 0:            # 递归结束条件
        fd(b)
        return
    
    draw_curve(b/3, n-1)  # 画线段①
    
    right(60)
    draw_curve(b/3, n-1)  # 画线段②
    
    left(120)
    draw_curve(b/3, n-1)  # 画线段③
    
    right(60)
    draw_curve(b/3, n-1)  # 画线段④
    return

def draw_snowflake(b, n):
    '''绘制科赫雪花'''
    for i in range(3):
        draw_curve(b, n)
        left(120)
    return

if __name__ == '__main__':
    '''绘制一个4阶的科赫雪花'''
    mode('logo')
    speed(0)
    pencolor('black')
    pensize(2)
    up()
    goto(-150, -90)
    seth(90)
    down()
    draw_snowflake(300, 4)
